Revise docs for removal of GCE Persistent Disk volume type

adityasamant25 opened the pull request:
<!--

 Hello!

 Remember to ADD A DESCRIPTION and delete this note before submitting
 your pull request. The description should explain what will change,
 and why.

 PLEASE title the FIRST commit appropriately, so that if you squash all
 your commits into one, the combined commit message makes sense.
 For overall help on editing and submitting pull requests, visit:
  https://kubernetes.io/docs/contribute/suggesting-improvements/

 Use the default base branch, “main”, if you're documenting existing
 features in the English localization.

 If you're working on a different localization (not English), see
 https://kubernetes.io/docs/contribute/new-content/overview/#choose-which-git-branch-to-use
 for advice.

 If you're documenting a feature that will be part of a future release, see
 https://kubernetes.io/docs/contribute/new-content/new-features/ for advice.

-->

Fixes https://github.com/kubernetes/website/issues/44359

This is a continuation of PR https://github.com/kubernetes/website/pull/44367, which I had to close as my local branch was corrupt.


I've handled the following changes:
[Types of Persistent Volumes](https://kubernetes.io/docs/concepts/storage/persistent-volumes/#types-of-persistent-volumes)
Added GCE Persistent Disk to the section of in-tree PV types that are no longer available.
Distinguished between volume plugins that are supported with CSI Migration vs plugins that are outright removed.
Used https://github.com/kubernetes/website/pull/44367#discussion_r1427097893 as reference

[Allowed Topologies](https://kubernetes.io/docs/concepts/storage/storage-classes/#allowed-topologies)
Modified the example in yaml to be based on a "sample" provisioner instead of kubernetes.io/gce-pd.
Also added some explanation about the allowedTopologies and matchLabelExpressions.


Suggestions and feedback are welcome.

Best Regards,
Aditya

sftim commented:
/sig storage

sftim reviewed (commented):
LGTM

adityasamant25 commented:
@xing-yang May I request I review from you on this PR?
As you were involved in the review for the previous PR #44367

adityasamant25 commented:
Please advise on the next step for this. I've awaiting a review/approval.
I'm available for any further changes that may need to be implemented.

msau42 commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -522,24 +522,29 @@ This means that support is still available but will be removed in a future Kuber
 
 * [`awsElasticBlockStore`](/docs/concepts/storage/volumes/#awselasticblockstore) - AWS Elastic Block Store (EBS)
-  (**not available** in v1.27)
+  (**not available** starting with v1.27, see [AWS EBS](https://github.com/kubernetes-sigs/aws-ebs-csi-driver) CSI driver)
```
Instead of "not available", I would say that these require the CSIMigration feature gate + CSI driver to be installed.

sftim replied:
```suggestion
  &mdash;this support is **only** available via a separate CSI driver for Kubernetes version 1.27 and onwards,
  as Kubernetes' core no longer provides built-in support. The CSI driver you need to install and configure is
  [AWS EBS](https://github.com/kubernetes-sigs/aws-ebs-csi-driver).
```

sftim commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -522,24 +522,29 @@ This means that support is still available but will be removed in a future Kuber
   (**deprecated** in v1.28)
 
-Older versions of Kubernetes also supported the following in-tree PersistentVolume types:
+The following in-tree PersistentVolume types are no longer available, but are supported through `CSIMigration` and have a corresponding CSI driver implemented.
```
Noted, however “in-tree“ doesn't sound quite right, if you need a CSI driver to use it.

msau42 replied:
maybe "built-in API"?

adityasamant25 replied:
I've changed it as follows in the latest commit.
`The following PersistentVolume types are supported through the CSIMigration feature gate and require the corresponding CSI driver to be installed in the cluster. Kubernetes offers built-in APIs to facilitate their use.`

sftim commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -522,24 +522,29 @@ This means that support is still available but will be removed in a future Kuber
+  (**not available** starting with v1.27, see [AWS EBS](https://github.com/kubernetes-sigs/aws-ebs-csi-driver) CSI driver)
 * [`azureDisk`](/docs/concepts/storage/volumes/#azuredisk) - Azure Disk
-  (**not available** in v1.27)
+  (**not available** starting with v1.27, see [Azure Disk](https://github.com/kubernetes-sigs/azuredisk-csi-driver) CSI driver)
```
```suggestion
  &mdash;this support is **only** available via a separate CSI driver for Kubernetes version 1.27 and onwards,
  as Kubernetes' core no longer provides built-in support. The CSI driver you need to install and configure is
  [Azure Disk](https://github.com/kubernetes-sigs/azuredisk-csi-driver).
```

sftim commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -522,24 +522,29 @@ This means that support is still available but will be removed in a future Kuber
 * [`cinder`](/docs/concepts/storage/volumes/#cinder) - Cinder (OpenStack block storage)
-  (**not available** in v1.26)
+  (**not available** starting with v1.26, see [OpenStack Cinder](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/cinder-csi-plugin/using-cinder-csi-plugin.md) CSI driver)
+* [`gcePersistentDisk`](/docs/concepts/storage/volumes/#gcepersistentdisk) - GCE Persistent Disk
+  (**not available** starting with v1.28, see [GCE Persistent Disk](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver) CSI driver)
```
```suggestion
  * [`gcePersistentDisk`](/docs/concepts/storage/volumes/#gcepersistentdisk) - GCE Persistent Disk -
    but this support is **only** available via a separate CSI driver for Kubernetes version 1.28 and onwards,
    as Kubernetes' core no longer provides built-in support. The CSI driver you need to install and configure is
    [GCE Persistent Disk](https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver) CSI-csi-driver).
```

sftim commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -522,24 +522,29 @@ This means that support is still available but will be removed in a future Kuber
+  (**not available** starting with v1.27, see [Azure Disk](https://github.com/kubernetes-sigs/azuredisk-csi-driver) CSI driver)
 * [`cinder`](/docs/concepts/storage/volumes/#cinder) - Cinder (OpenStack block storage)
-  (**not available** in v1.26)
+  (**not available** starting with v1.26, see [OpenStack Cinder](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/cinder-csi-plugin/using-cinder-csi-plugin.md) CSI driver)
```
```suggestion
    &mdash; but for Kubernetes version 1.26 and onwards, this volume type is **only** available via a separate CSI driver 
    as Kubernetes' core no longer provides built-in support. The CSI driver you need to install and configure is
    [OpenStack Cinder](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/cinder-csi-plugin/).
```

sftim commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -522,24 +522,29 @@ This means that support is still available but will be removed in a future Kuber
 * [`rbd`](/docs/concepts/storage/volumes/#rbd) - Rados Block Device (RBD) volume
```
For several types of storage we're now talking about volume types that are already only provided through CSI drivers. The migrations should already have happened, right?

msau42 replied:
When we talk about the CSI migration feature, that means users can continue using the built-in type in the API, but in the backend we redirect all the operations to the CSI driver. This means users can continue using the same objects and APIs.

There are other volume types that we deprecated without a CSI migration implementation. That means any existing objects will no longer be functional, and users have to create new objects that directly reference the CSI volume type.

adityasamant25 replied:
To get the list of volume types that are (or will be) supported by CSI migration, I am referring to https://github.com/kubernetes/website/pull/44367#discussion_r1427097893

As per that list,
azureFile - planned for 1.30, so we can add it to the list of CSI migration after 1.30 is GA. I propose no change for now.
flexVolume - Not in the CSI migration plan. No change needed in the documentation.
portworxVolume - no information available about in-tree plugin removal. So no change needed in the documentation.
vsphereVolume - planned for 1.30, so we can add it to the list of CSI migration after 1.30 is GA. I propose no change for now.
cephfs - Not in the CSI migration plan. No change needed in the documentation.
rbd - Not in the CSI migration plan. No change needed in the documentation.

Please keep me honest here.

sftim commented on content/en/docs/concepts/storage/storage-classes.md:
````diff
@@ -242,18 +242,20 @@ apiVersion: storage.k8s.io/v1
+    - zone1
+    - zone2
 ```
 
````
```suggestion

Here's an outline of some important fields within the [StorageClass](/docs/reference/kubernetes-api/config-and-storage-resources/storage-class-v1/) API:

```

sftim commented on content/en/docs/concepts/storage/storage-classes.md:
```diff
@@ -242,18 +242,20 @@ apiVersion: storage.k8s.io/v1
 
+- `provisioner`: See [`Provisioner`](/docs/concepts/storage/storage-classes/#provisioner) for details.
+- `allowedTopologies`: Defines the constraints for the allowed topologies. In this example, it's based on zone.
+- `matchLabelExpressions`: Specify label expressions to enable a node to be considered part of the allowed topology. For example, the node must have a label `topology.kubernetes.io/zone` with a value of either `zone1` or `zone2`.
```
How about:
```suggestion
- `matchLabelExpressions`: Specify selectors to enable a node to be considered
  part of the allowed topology.
  In the example manifest, PersistentVolumes for this StorageClass must have a label
  `topology.kubernetes.io/zone`,   with a label value of either `zone1` or `zone2`.
```

sftim commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -522,24 +522,29 @@ This means that support is still available but will be removed in a future Kuber
+
 * `photonPersistentDisk` - Photon controller persistent disk.
-  (**not available** starting v1.15)
+  (**not available** starting with v1.15)
```
We could rewrite this and comment that there's no CSI driver for Photon available from Kubernetes.

adityasamant25 replied:
There is no CSI driver from Kubernetes for all the types in this section, not only Photon, right?
Should we add a statement such as `These volume types do not have a CSIMigration plan and are discontinued in newer Kubernetes releases.` ?

xing-yang replied:
In the following page, we can find CSI drivers for Quobyte and StorageOS. However, since this is updated by maintainers of individual CSI drivers, the info can be stale as well. It's probably better not to mention whether there's a CSI driver or not as it could also change.
https://kubernetes-csi.github.io/docs/drivers.html

sftim replied:
I think we should say when there are _in project_ drivers available. Readers would want to know.

adityasamant25 replied:
Thanks @xing-yang and @sftim. I understand it now. I have made an attempt to add this information in my latest commit.

adityasamant25 commented:
All the review comments received till now have been implemented.

dipesh-rawat commented:
Hello @kubernetes/sig-storage-pr-reviews, we would appreciate your feedback to ensure the accuracy of these proposed changes. Thanks!

jsafrane commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -522,24 +522,33 @@ This means that support is still available but will be removed in a future Kuber
-Older versions of Kubernetes also supported the following in-tree PersistentVolume types:
+The following PersistentVolume types are supported through the `CSIMigration` feature gate and require the corresponding CSI driver to be installed in the cluster. Kubernetes offers built-in APIs to facilitate their use.
 
 * [`awsElasticBlockStore`](/docs/concepts/storage/volumes/#awselasticblockstore) - AWS Elastic Block Store (EBS)
```
There are these volume plugins / corresponding CSI drivers missing:

* AzureFile CSI migration went GA in 1.26
* vSphere CSI migration went GA in 1.26

IMO it would make more sense to have a table with in-tree volume plugin name, link to the CSI driver, and  Kubernetes version, with generic `this volume type is **only** available via a separate CSI driver 
  as Kubernetes' core no longer provides built-in support.` below it.

sftim replied:
> IMO it would make more sense to have a table with in-tree volume plugin name, link to the CSI driver, and Kubernetes version, with generic `this volume type is **only** available via a separate CSI driver as Kubernetes' core no longer provides built-in support.` below it.

PRs very welcome (from any contributor)

adityasamant25 replied:
@sftim @jsafrane Unfortunately, I do not have the capacity currently to make this formatting change.
I request to not include this formatting change in the scope of this PR. If needed it can be opened as an enhancement issue and a contributor who would have the capacity could pick it up.

xing-yang replied:
As commented early by @jsafrane , AzureFile and vSphere are missing.

jsafrane commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -522,24 +522,33 @@ This means that support is still available but will be removed in a future Kuber
 
 * [`awsElasticBlockStore`](/docs/concepts/storage/volumes/#awselasticblockstore) - AWS Elastic Block Store (EBS)
-  (**not available** in v1.27)
+  &mdash; but for Kubernetes version 1.27 and onwards, this volume type is **only** available via a separate CSI driver 
```
AWS EBS CSI migration went [GA in 1.25](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.25.md#feature-14)

jsafrane replied:
No, the CSI driver was really mandatory in 1.25. A Kubernetes cluster is not able to provision/delete or mount/unmount in-tree `awsElasticBlockStore` volumes without it. The only exception is attaching and detaching them to/from kubelets < 1.25, that's why we removed it in 1.27.

This applies to all plugins below.

adityasamant25 replied:
Updated the version in the latest commit.

jsafrane commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -522,24 +522,33 @@ This means that support is still available but will be removed in a future Kuber
-  (**not available** starting v1.25)
-* `storageos` - StorageOS volume.
-  (**not available** starting v1.25)
+    &mdash; but for Kubernetes version 1.26 and onwards, this volume type is **only** available via a separate CSI driver 
```
CSI migration is GA since 1.24

adityasamant25 replied:
The documentation is capturing the version since the use of the CSI driver was mandatory. So 1.24 is correct, right?

adityasamant25 replied:
Updated the version in the latest commit.

jsafrane commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -522,24 +522,33 @@ This means that support is still available but will be removed in a future Kuber
+    as Kubernetes' core no longer provides built-in support. The Kubernetes project provides an equivalent
+    CSI driver,
+    [OpenStack Cinder](https://github.com/kubernetes/cloud-provider-openstack/blob/master/docs/cinder-csi-plugin/).
+* [`gcePersistentDisk`](/docs/concepts/storage/volumes/#gcepersistentdisk) - GCE Persistent Disk &mdash; but for Kubernetes version 1.28 and onwards, this volume type is **only** available via a separate CSI driver 
```
GCE PD CSI migration is GA since 1.25

adityasamant25 replied:
The documentation is capturing the version since the use of the CSI driver was mandatory. So 1.28 is correct, right?

adityasamant25 replied:
Updated the version in the latest commit.

jsafrane commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -522,24 +522,33 @@ This means that support is still available but will be removed in a future Kuber
+  CSI driver, [AWS EBS](https://github.com/kubernetes-sigs/aws-ebs-csi-driver).
 * [`azureDisk`](/docs/concepts/storage/volumes/#azuredisk) - Azure Disk
-  (**not available** in v1.27)
+  &mdash; but for Kubernetes version 1.27 and onwards, this volume type is **only** available via a separate CSI driver
```
AzureDisk CSI migration is [GA since 1.24](https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.24.md#feature-17)

I.e. 1.24 already required the CSI driver  to be installed. In 1.27 we only removed the in-tree code for AWS EBS, but it has been (mostly) unused for 2 releases already.

adityasamant25 replied:
The documentation is capturing the version since the use of the CSI driver was mandatory. So 1.27 is correct, right?

adityasamant25 replied:
Updated the version in the latest commit.

sftim commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -534,18 +534,33 @@ please install corresponding CSI drivers.
+
 Older versions of Kubernetes also supported the following in-tree PersistentVolume types:
 
-* `photonPersistentDisk` - Photon controller persistent disk.
-  (**not available** starting v1.15)
-* `scaleIO` - ScaleIO volume.
-  (**not available** starting v1.21)
-* `flocker` - Flocker storage.
-  (**not available** starting v1.25)
-* `quobyte` - Quobyte volume.
-  (**not available** starting v1.25)
-* `storageos` - StorageOS volume.
-  (**not available** starting v1.25)
```
Should we list GCE PD here, rather than in the list of available drivers?

sftim commented on content/en/docs/concepts/storage/persistent-volumes.md:
```diff
@@ -534,18 +534,33 @@ please install corresponding CSI drivers.
 * [`vsphereVolume`](/docs/concepts/storage/volumes/#vspherevolume) - vSphere VMDK volume
   (**migration on by default** starting v1.25)
 
+The following PersistentVolume types are supported through the `CSIMigration` feature gate and require the corresponding CSI driver to be installed in the cluster. Kubernetes offers built-in APIs to facilitate their use.
```
Is that right? AIUI there is no longer a `CSIMigration` feature gate in current Kubernetes.

adityasamant25 replied:
@sftim, this was worded as per the advice received from @msau42.
@msau42, request you to kindly review if these sentences are appropriate.

jsafrane replied:
I agree with @sftim, CSIMigration feature gate is now removed.

xing-yang replied:
This comment is not addressed yet.  The core CSIMigration feature is GA and the feature gate is removed.
Suggest change to:
The following PersistentVolume types support CSIMigration and require the corresponding CSI driver to be installed in the cluster. Kubernetes offers built-in APIs to facilitate their use.

adityasamant25 replied:
@xing-yang This has been incorporated in the latest commit.

divya-mohan0209 commented:
@adityasamant25 : Gentle reminder on this PR.

adityasamant25 commented:
> @adityasamant25 : Gentle reminder on this PR.

Thanks @divya-mohan0209. I have pushed another commit addressing the comments that needed to be fixed.

I have replied to the other open comments.

divya-mohan0209 commented:
Ping @kubernetes/sig-storage-pr-reviews for reviewing the technical accuracy.

T-Lakshmi commented:
@adityasamant25,
PR conflicting with current changes, Please rebase the PR towards main.

divya-mohan0209 commented:
@adityasamant25 : Please could you address the rebase?

@xing-yang / @kubernetes/sig-storage-pr-reviews : PTAL for technical accuracy

xing-yang commented:
@divya-mohan0209 There are still a few comments that are not addressed yet.  Thanks.

adityasamant25 commented:
@T-Lakshmi @divya-mohan0209 
I apologize for not handling the rebase earlier. I was caught up in other work.

I have addressed the rebase now.

I do not have the bandwidth to continue working on this PR, so I would appreciate if another contributor would like to own the PR until it is merged.

tengqm commented:
/approve

nate-double-u commented:
Thanks for all your work on this @adityasamant25 (and to all the reviewers)! It looks like @120EE0980 has taken the reigns to move this work forward in https://github.com/kubernetes/website/pull/49747.

I'm going to close this PR in favour of that one. Please feel free to reopen it if I'm wrong about that.

/close
